### ISSUE TITLE
[ENH] Use the latest protocol 5 for serialization of objects using pickle

### ISSUE THREAD

**Issue by achieveordie** [sktime/sktime#5541]
**Is your feature request related to a problem? Please describe.**
The current default version is [4](https://docs.python.org/3/library/pickle.html#pickle.DEFAULT_PROTOCOL). Protocol 5 suits our needs better where we might be storing big chunks of data like a numpy array. Here's an excerpt from [PEP 574](https://peps.python.org/pep-0574/) which brought in this change:

> Nowadays the pickle protocol sees a growing use in applications where most of the data isn’t ever persisted to disk (or, when it is, it uses a portable format instead of Python-specific). Instead, pickle is being used to transmit data and commands from one process to another, either on the same machine or on multiple machines. Those applications will sometimes deal with very large data (such as Numpy arrays or Pandas dataframes) that need to be transferred around. For those applications, pickle is currently wasteful as it imposes spurious memory copies of the data being serialized.
...
This PEP aims to make pickle usable in a way where large data is handled as a separate stream of zero-copy buffers, letting the application handle those buffers optimally.

**Describe the solution you'd like**
Since we only support Python 3.8 and above, we can introduce this protocol (by simply mentioning `protocol=5` during the dump) to gain some performance compared to the default version.

**Describe alternatives you've considered**
We leave things as they are.

**Additional context**

**Comment by fkiraly** [comment]
Would we then end up with sth similar to the parallelization ackend, where we specify serialization format plus parameters?

**Comment by achieveordie** [comment]
When I opened this issue, I was of the opinion of not providing any parameter to the end user for two reasons:
- There are no actual cons of using Protocol 5 instead of 4 (perhaps the default is still 4 to support a wider range of Python versions).
- The end user is less likely to care about which protocol is being used to serialize their data.

The only situation where Protocol 5 is less desirable is when a user is using a legacy version of Python with sktime and wants to serialize their estimators.

**Comment by fkiraly** [comment]
> The only situation where Protocol 5 is less desirable is when a user is using a legacy version of Python with sktime and wants to serialize their estimators.

Yes, and we have deprecation/change as well as support for earlier serialized models to think of. The model formats should be downwards compatible, serialized models typically have a much longer lifetime than boilerplate code.

**Comment by achieveordie** [comment]
In that case, we should provide `pickle_protocol` as a parameter with the default of 5. It can be changed to 4 in case someone is using legacy versions. What do you think?

**Comment by fkiraly** [comment]
>  It can be changed to 4 in case someone is using legacy versions. What do you think?

Should it perhaps be a `serialization_params` `dict`, to allow passing similar params to all backends?

**Comment by achieveordie** [comment]
Looking through `cloudpickle`'s codebase I [found](https://github.com/cloudpipe/cloudpickle/blob/master/cloudpickle/cloudpickle.py#L1452) that by default they already use the highest pickle protocol version for dumping. Probably this is why `cloudpickle` is faster than `pickle`.

**Comment by achieveordie** [comment]
> Should it perhaps be a serialization_params dict, to allow passing similar params to all backends?

The two parameters that our current backends share are `protocol` and `buffer_callback`. `protocol` shouldn't warrant any special testing but we might need to test the behavior of `save` if `buffer_callback` is non-null.

**Comment by fkiraly** [comment]
> The two parameters that our current backends share

Yes, but I'm thinking of upwards compatibility with regards to additional backends we may like to add in the future.

Like as we have learnt in the case of parallelization, different backends will have different sets of parameters, and I only see that working if you have a param dict. Of course you could catch it as kwargs if there's no likely ambiguity - but there unfortunately may be, as backends can have their own backends https://github.com/sktime/sktime/pull/5537

FYI @SmirnGregHM due to the related parallelization backend discussion, in case you have any remarks from the user perspective.
